### ISSUE TITLE
In the AST, model activities more cleanly 

### ISSUE THREAD

**Issue by karl** [sverweij/state-machine-cat#22]
### Expected Behavior

I'm trying to create a tool to create svg visualisations of xstate state machines.

To do this I'm attempting to convert xstate machine descriptions to State Machine Cat json, which I can then feed into State Machine Cat to generate an svg.

Alongside entry and exit triggers, xstate supports the idea of activities. At the moment State Machine Cat json has a hard coded list of acceptable trigger types, which only include `entry` and `exit`.

https://github.com/sverweij/state-machine-cat/blob/develop/src/parse/smcat-ast.schema.json#L29

Would it be possible to add `activity` (or some equivalent) to that list?

### Steps to Reproduce (for bugs)

I can whip up a minimal example of the problem if that is useful!

**Comment by sverweij** [comment]
@karl thanks for raising this

state-machine-cat interprets everything after a `:` as 'activities' (see _basic activities_ below). The internal representation is a bit hacky, though. It suffices for graphical rendering, but is imperfect for conceptual interpretation (see _more than one activity_ and _mixed activities/ entry/ exit_).

So I'll try & make it less hacky - which should help in resolving this issue (see below - but I'm open to feedback).

As a side note - I want to make state-machine-cat better interoperate with xstate anyway. If you bump into something else I'll be interested.

## Representation of _activities_ - sketch of a desired situation

```
play: 
   entry/make some noise
   blink a light
   do something funny
   exit/click;
```

We _can_ introduce a keyword for activities as well, but it'll be equivalent to not having it e.g.

```
play: 
   entry/make some noise
   activity/blink a light
   activity/do something funny
   exit/click;
```
which might be helpful for readability.

- each line after the `:` that contains `entry/` or `exit/` is an entry (/ exit) trigger
- other separate lines become separate activities
- These activities go into the `activities` attribute - which will become an array of strings, instead of a string:

```json
{
    "states": [
        {
            "name": "play",
            "type": "regular",
            "activities": [
                "blink a light",
                "do something funny"
            ],
            "triggers": [
                {
                    "type": "entry",
                    "body": "make some noise"
                },
                {
                    "type": "exit",
                    "body": "click"
                }
            ]
        }
    ]
}
```

## Representation of _activities_ -  current situation
### basic activities - 👍 
```
play: make some noise;
```

```json
{
    "states": [
        {
            "name": "play",
            "type": "regular",
            "activities": "make some noise"
        }
    ]
}
```

<img width="152" alt="activities" src="https://user-images.githubusercontent.com/4822597/45972353-cb3a5000-c03b-11e8-816a-6b8461ac549c.png">

(On translation to SCXML _activities_ that didn't have entry or exit in them translate to _onentry_ as that's the closest thing in scxml that works)
```scxml
<?xml version="1.0" encoding="UTF-8"?>
<scxml xmlns="http://www.w3.org/2005/07/scxml" version="1.0">
    <state id="play">
        <onentry>make some noise</onentry>
    </state>
</scxml>
```

### more than one activity - 🤔
```
play: 
   make some noise
   blink a light
   click;
```


```json
{
    "states": [
        {
            "name": "play",
            "type": "regular",
            "activities": "make some noise\n   blink a light\n   click"
        }
    ]
}
```

### mixed activities/ entry/ exit - 🤒 
```
play: 
   entry/make some noise
   blink a light
   exit/click;
```

```json
{
    "states": [
        {
            "name": "play",
            "type": "regular",
            "activities": "entry/make some noise\n   blink a light\n   exit/click",
            "triggers": [
                {
                    "type": "entry",
                    "body": "make some noise"
                },
                {
                    "type": "exit",
                    "body": "click"
                }
            ]
        }
    ]
}
```

**Comment by sverweij** [comment]
Version 3.0.4-beta-0 on npm contains the first crude implementation of the above proposal (which needs polishing (see the todo list in the PR), but should do what is promised on the tin).

Let me know if that'll work for you

```json
{
    "states": [
        {
            "name": "A",
            "type": "regular",
            "activities": [
                "throughout X"
            ],
            "triggers": [
                {
                    "type": "entry",
                    "body": "S"
                },
                {
                    "type": "exit",
                    "body": "Beta and T"
                }
            ]
        }
    ]
}
```

<img width="137" alt="screen shot 2018-09-25 at 19 53 15" src="https://user-images.githubusercontent.com/4822597/46032757-c4274680-c0fc-11e8-931f-19181a4cf594.png">

**Comment by karl** [comment]
Thanks for the quick turn around @sverweij, amazing stuff!

Hopefully I'll have a chance to play around with this over the next few days and give some feedback. But the main concepts and ideas seem great.

**Comment by karl** [comment]
I've had a little play around with this and it looks good!

It's allowed me to remove a hack I'd put in place to visualise activities which is great. You can see a (messy) PR with my changes here: https://github.com/karl/redux-saga-state-machine/pull/145

Would you consider having State Machine Cat order the triggers and activities in the SVG in the order they would be run (entry, activities, exit)?

**Comment by sverweij** [comment]
@ karl/redux-saga-state-machine#145: definitely an improvement :-)

@ ordering triggers and activities in the order they would be run: it was one of the options I considered, but I prefer to have smc render in the order the user entered it - in that way the user stays in control of (that part of) the output; he or she might have reasons to deviate from what seems logical to us. If even David Harel does so (see figure 37 from his original paper), who am I to come in between ...:

<img width="141" alt="screen shot 2018-09-29 at 16 49 55" src="https://user-images.githubusercontent.com/4822597/46247179-df010000-c407-11e8-907b-bd501ceb4370.png">

**Comment by sverweij** [comment]
I've been struggling with keeping the ordering as entered, but the code became really messy. So I had to get back to base 1: the model wasn't good enough yet. `entry` and `exit`'s are similar to `activities` - but in the model they weren't.

I took a look at the Harel's original state chart paper - and noticed he calls all three of them _actions_ (similar to the same concept on transitions). So I decided to 

- rename `triggers` to `actions`
- move activities under the actions with `activity` as `type`

... which results in this AST for the above sample:
```json
{
    "states": [
        {
            "name": "play",
            "type": "regular",
            "actions": [
                {
                    "type": "entry",
                    "body": "make some noise"
                },
                {
                    "type": "activity",
                    "body": "blink a light"
                },
                {
                    "type": "activity",
                    "body": "do something funny"
                },
                {
                    "type": "exit",
                    "body": "click"
                }
            ]
        }
    ]
}
```

Published as `3.0.4-beta-2 ` on npm.

(I've taken the liberty to update redux-saga-state-machine/xstate-to-svg and retest it, and it seems to work OK, provided `jest -u` is run to update the snapshot so it expects the new (more correct) order of actions is  - diff [here](https://gist.github.com/sverweij/238ac58539cecec834fab8e14f1e13f2/revisions)

**Comment by karl** [comment]
Awesome!

That is exactly how I was modelling it in my head, so I’m very happy with
that change 😀
